test/NuGet.Core.Tests/NuGet.PackageManagement.Test/DependencyGraphSpecTests.cs

```diff
@@ -0,0 +1,48 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using NuGet.ProjectModel;
+using Xunit;
+
+namespace NuGet.Test
+{
+    public class DependencyGraphSpecTests
+    {
+        [Fact]
+        public void DependencyGraphSpecTests_WithPackageSpecs()
+        {
+            // Arrange
+            var packageSpecA = new PackageSpec();
+            packageSpecA.Title = "A";
+            packageSpecA.RestoreMetadata = new ProjectRestoreMetadata() { ProjectUniqueName = "a", CentralPackageVersionsEnabled = false };
+            var dgSpec = new DependencyGraphSpec();
+            dgSpec.AddRestore("a");
+            dgSpec.AddProject(packageSpecA);
+            var packageSpecB = new PackageSpec();
+            packageSpecB.Title = "B";
+            packageSpecB.RestoreMetadata = new ProjectRestoreMetadata()
+            {
+                ProjectUniqueName = "BBB"
+            };
+            var packageSpecC = new PackageSpec();
+            packageSpecC.Title = "C";
+            packageSpecC.RestoreMetadata = new ProjectRestoreMetadata()
+            {
+                ProjectUniqueName = "CCC"
+            };
+            var newNugetPackageSpecs = new List<PackageSpec>()
+            {
+                packageSpecB,
+                packageSpecC
+            };
+
+            // Act
+            dgSpec = dgSpec.WithPackageSpecs(newNugetPackageSpecs);
+
+            // Assert
+            Assert.Equal(dgSpec.Projects.Count, 3);
```

Comment:
Yes, it does just addition. My new method is just override/extension of [`WithReplacedSpec`](https://github.com/NuGet/NuGet.Client/blob/156230a57ee2c7f425dec377689b8d16d31b480b/src/NuGet.Core/NuGet.ProjectModel/DependencyGraphSpec.cs#L473), which in reality it just does addition. I already renamed my new method to `WithPackageSpecs`.

Reply:
The way it's currently implemented, does it do anything?

It's expected that the project we are modifying already exists, so calling `AddProject` won't replace it.

Reply:
Yes, actually it just add projects but not replace it.
I just created as  extension for [`WithReplacedSpec`](https://github.com/NuGet/NuGet.Client/blob/156230a57ee2c7f425dec377689b8d16d31b480b/src/NuGet.Core/NuGet.ProjectModel/DependencyGraphSpec.cs#L473) which only adds just 1 projects at the time, but mine add many.

Reply:
I took the liberty to quickly write a test for that: 

```cs
[Fact]
        public void WithReplacedPackageSpecs_WithASinglePackageSpec_Succeeds()
        {
            // Arrange
            var packageSpecA = new PackageSpec
            {
                Title = "A",
                RestoreMetadata = new ProjectRestoreMetadata()
                {
                    ProjectUniqueName = "a", CentralPackageVersionsEnabled = false
                }
            };
            var packageSpecB = new PackageSpec
            {
                Title = "B",
                RestoreMetadata = new ProjectRestoreMetadata()
                {
                    ProjectUniqueName = "BBB"
                }
            };
            var packageSpecC = new PackageSpec
            {
                Title = "C",
                RestoreMetadata = new ProjectRestoreMetadata()
                {
                    ProjectUniqueName = "CCC"
                }
            };
            var dgSpec = new DependencyGraphSpec();
            dgSpec.AddProject(packageSpecA);
            dgSpec.AddProject(packageSpecB);
            dgSpec.AddProject(packageSpecC);
            dgSpec.AddRestore(packageSpecA.RestoreMetadata.ProjectUniqueName);
            dgSpec.AddRestore(packageSpecB.RestoreMetadata.ProjectUniqueName);
            dgSpec.AddRestore(packageSpecC.RestoreMetadata.ProjectUniqueName);

            // Create an updated packageSpecA
            var updatedPackageA = packageSpecA.Clone();
            updatedPackageA.RestoreMetadata.ConfigFilePaths.Add("/samplePath");
            var newNugetPackageSpecs = new List<PackageSpec>()
            {
                updatedPackageA
            };

            // Preconditions
            dgSpec.Projects.Should().HaveCount(3);
            dgSpec.Restore.Should().HaveCount(3);

            // Act
            var dgSpecWithReplacedPackageA = dgSpec.WithPackageSpecs(newNugetPackageSpecs);

            // Assert
            dgSpecWithReplacedPackageA.Projects.Should().HaveCount(3);
            dgSpecWithReplacedPackageA.Restore.Should().HaveCount(1);

            var packageSpecInAFromDgSpec = dgSpecWithReplacedPackageA.Projects.Single(e => e.Title.Equals("A"));
            packageSpecInAFromDgSpec.Should().Be(updatedPackageA);
            dgSpecWithReplacedPackageA.Restore.Single().Should().Be(updatedPackageA.RestoreMetadata.ProjectUniqueName);
        }
```

I think the name should be `WithReplacedPackageSpecs`. It'll match the behavior of `WithReplacedPackageSpec`. 
It returns a complete new dg spec version, where the package specs in this case are equivalent to the original one, unless they are part of the list in the arguments.

The reason why I insisted on this test is because testing only one happy scenario is never enough. 
Unit tests should be more detailed and consider more scenario (both positive and negative) so that the callers are less likely to run into unexpected situations. 

Some additional examples:

* WithReplacedPackageSpecs, when the package spec argument has only one value, should run do the same thing as WithReplacePackageSpec.
* WithReplacedPackageSpec, with a package spec that never a part of the original dg spec etc.